File: src/main/java/me/taubsie/dungeonhub/application/listener/MessageListener.java
package me.taubsie.dungeonhub.application.listener;

import me.taubsie.dungeonhub.application.classes.ServerProperty;
import me.taubsie.dungeonhub.application.config.ConfigProperty;
import me.taubsie.dungeonhub.application.connection.MojangConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.ContentConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.DiscordUserConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.QueueConnection;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.ScoreConnection;
import me.taubsie.dungeonhub.application.enums.EmbedColor;
import me.taubsie.dungeonhub.application.enums.IdList;
import me.taubsie.dungeonhub.application.exceptions.FailedToLoadEmbedException;
import me.taubsie.dungeonhub.application.exceptions.PlayerNotFoundException;
import me.taubsie.dungeonhub.application.service.ApplicationService;
import me.taubsie.dungeonhub.application.service.LeaderboardService;
import me.taubsie.dungeonhub.common.DungeonHubService;
import me.taubsie.dungeonhub.common.enums.QueueStep;
import me.taubsie.dungeonhub.common.enums.ScoreType;
import me.taubsie.dungeonhub.common.model.carry_queue.CarryQueueModel;
import me.taubsie.dungeonhub.common.model.carry_queue.CarryQueueUpdateModel;
import me.taubsie.dungeonhub.common.model.discord_user.DiscordUserModel;
import me.taubsie.dungeonhub.common.model.score.LoggedCarryModel;
import me.taubsie.dungeonhub.common.model.score.ScoreModel;
import org.javacord.api.entity.channel.PrivateChannel;
import org.javacord.api.entity.channel.ServerTextChannel;
import org.javacord.api.entity.channel.TextChannel;
import org.javacord.api.entity.message.Message;
import org.javacord.api.entity.message.MessageAttachment;
import org.javacord.api.entity.message.MessageUpdater;
import org.javacord.api.entity.message.component.*;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.entity.server.Server;
import org.javacord.api.entity.user.User;
import org.javacord.api.event.message.CertainMessageEvent;
import org.javacord.api.event.message.MessageCreateEvent;
import org.javacord.api.event.message.MessageEditEvent;
import org.javacord.api.listener.message.MessageCreateListener;
import org.javacord.api.listener.message.MessageEditListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.CompletionException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author Taubsie
 * @since 1.0.0
 */
@Listener
public class MessageListener implements MessageCreateListener, MessageEditListener {
    private static final Logger logger = LoggerFactory.getLogger(MessageListener.class);

    private static final long APPROVE_AMOUNT_THRESHOLD = 8;
    private static final long APPROVE_SCORE_THRESHOLD = 30;

    private static final Pattern CHANNEL_FROM_TRANSCRIPT = Pattern.compile("^\\s*Channel: [^(]*\\((?<channel>\\d*)\\)");

    @Override
    public void onMessageCreate(MessageCreateEvent messageCreateEvent) {
        addReactionToPets(messageCreateEvent);

        logTicket(messageCreateEvent);

        loadSkycryptFromTicket(messageCreateEvent);
    }

    @Override
    public void onMessageEdit(MessageEditEvent messageEditEvent) {
        logTicket(messageEditEvent);
    }

    private void addReactionToPets(MessageCreateEvent messageCreateEvent) {
        if (messageCreateEvent.isServerMessage() && messageCreateEvent.getServer().isPresent()) {
            long serverId = messageCreateEvent.getServer().get().getId();
            long channelId = messageCreateEvent.getChannel().getId();
            if ((serverId == 1023684107877761196L && channelId == 1220895875102937098L)
                    || (serverId == 693263712626278553L && channelId == 1219427157655289908L)) {
                if (messageCreateEvent.getMessageAttachments().isEmpty()) {
                    if (messageCreateEvent.getMessageAuthor().isRegularUser() && messageCreateEvent.getMessage().getEmbeds().stream().flatMap(embed -> embed.getThumbnail().stream()).findFirst().isEmpty()) {
                        return;
                    }
                }

                String emoji = getRandomEmoji();

                try {
                    messageCreateEvent.addReactionToMessage(emoji).join();
                }
                catch (CompletionException ignored) {
                    // ignored, just don't add a reaction then
                }
            }
        }
    }

    private String[] getEmojiPool() {
        return new String[]{
                "Woah:1220111116651204608",
                "woah:1220111081150615572",
                "girlwow:1220111157742800956",
                "catelove:1204407157848678430",
                "ZTcool:1204406493353353256",
                "pepega:697756021048868894",
                "smikecate:1204406375791333426",
                "poggorfish:694270485613117480"
        };
    }

    private String getRandomEmoji() {
        double bound = 101.0;

        String[] emojiPool = getEmojiPool();

        int random = new Random().nextInt((int) bound);

        String emoji = emojiPool[(int) (random * (emojiPool.length / bound))];

        if (random == 100) {
            emoji = "swag:708383726370947132";
        }

        return emoji;
    }

    private void loadSkycryptFromTicket(MessageCreateEvent messageCreateEvent) {
        Optional<Server> server = messageCreateEvent.getServer();
        Optional<ServerTextChannel> channel = messageCreateEvent.getMessage().getServerTextChannel();

        if (!messageCreateEvent.isServerMessage()
                || channel.isEmpty()
                || server.isEmpty()
                || !(server.get().getId() == IdList.SERVER.getId()
                || server.get().getId() == IdList.SERVER.getTestId())) {
            return;
        }

        try {
            if (channel.get().getMessages(5).join().size() != 1) {
                return;
            }

            Message firstMessage =
                    channel.get().getMessagesAsStream().reduce((message, message2) -> message2).orElse(null);

            if (firstMessage == null) {
                return;
            }

            List<User> mentionedUsers = firstMessage.getMentionedUsers();

            if (mentionedUsers.size() != 1) {
                return;
            }

            User user = mentionedUsers.get(0);

            String[] lines = firstMessage.getContent().split("\n");

            if (lines.length < 2) {
                return;
            }

            Optional<String> ignOptional = DiscordUserConnection.getInstance()
                    .getLinkedById(user.getId())
                    .map(DiscordUserModel::getMinecraftId)
                    .map(uuid -> MojangConnection.getInstance().getNameByUUID(uuid))
                    .or(() -> Arrays.stream(lines)
                            .filter(s -> s.startsWith("IGN: "))
                            .findFirst())
                    .or(() -> user.getNickname(server.get()));

            if (ignOptional.isEmpty()) {
                return;
            }

            String ign = ignOptional.get()
                    .replace("IGN: ", "")
                    .replaceAll("❮(\\S*)❯", "")
                    .replace("★", "")
                    .replace("✦", "")
                    .replace("✶", "")
                    .replace("✽", "")
                    .replace("❊", "")
                    .strip();

            sendPlayerDataEmbed(ign, channel.get());
        }
        catch (CompletionException ignored) {
            //this just happens when the execution takes so long that the channel gets deleted
            //sending an error then wouldn't be needed
        }
    }

    private Button getSkyCryptButton(String ign) {
        return new ButtonBuilder().setStyle(ButtonStyle.LINK)
                .setUrl(ConfigProperty.SKYCRYPT_API_URL + "stats/" + ign)
                .setLabel("SkyCrypt")
                .build();
    }

    //TODO threads threads threads
    private void sendPlayerDataEmbed(String ign, ServerTextChannel channel) {
        EmbedBuilder playerDataEmbed;
        try {
            playerDataEmbed = ApplicationService.getInstance().getPlayerDataEmbed(ign, null);

            channel.sendMessage(playerDataEmbed,
                    new ActionRowBuilder().addComponents(
                            getSkyCryptButton(ign)
                    ).build());
        }
        catch (PlayerNotFoundException playerNotFoundException) {
            playerDataEmbed = ApplicationService.getInstance().getErrorEmbed(playerNotFoundException);

            //TODO load scammer data from discord?

            channel.sendMessage(playerDataEmbed);
        }
        catch (FailedToLoadEmbedException failedToLoadEmbedException) {
            playerDataEmbed = failedToLoadEmbedException.getEmbed();

            channel.sendMessage(playerDataEmbed,
                            new ActionRowBuilder().addComponents(
                                    getSkyCryptButton(ign),
                                    new ButtonBuilder().setStyle(ButtonStyle.SECONDARY)
                                            .setLabel("Reload")
                                            .setCustomId("reload_playerdata")
                                            .build()
                            ).build())
                    .thenAccept(message -> message.addButtonClickListener(event ->
                            event.getButtonInteractionWithCustomId("reload_playerdata")
                                    .ifPresent(buttonInteraction -> {
                                        buttonInteraction.createOriginalMessageUpdater()
                                                .removeAllComponents()
                                                .removeAllEmbeds()
                                                .addEmbed(ApplicationService.getInstance()
                                                        .getEmbed()
                                                        .setDescription("Loading..."))
                                                .update().join();

                                        //TODO maybe also update username to use?

                                        MessageUpdater updater = message.createUpdater();

                                        try {
                                            EmbedBuilder embed =
                                                    ApplicationService.getInstance().getPlayerDataEmbed(ign, null);

                                            updater.removeAllEmbeds()
                                                    .addEmbed(embed);

                                            updater.removeAllComponents()
                                                    .addComponents(new ActionRowBuilder().addComponents(
                                                            getSkyCryptButton(ign)
                                                    ).build());
                                        }
                                        catch (FailedToLoadEmbedException failedToLoadAgain) {
                                            updater.removeAllEmbeds()
                                                    .addEmbed(failedToLoadAgain.getEmbed());

                                            updater.removeAllComponents()
                                                    .addComponents(new ActionRowBuilder().addComponents(
                                                            getSkyCryptButton(ign),
                                                            new ButtonBuilder().setStyle(ButtonStyle.SECONDARY)
                                                                    .setLabel("Reload")
                                                                    .setCustomId("reload_playerdata")
                                                                    .build()
                                                    ).build());
                                        }

                                        updater.applyChanges();
                                    })
                    ));
        }
    }

    //TODO reduce complexity
    private void logTicket(CertainMessageEvent messageEvent) {
        Server server = messageEvent.getServer().orElse(null);

        if (!messageEvent.isServerMessage() || server == null) {
            return;
        }

        if (ServerProperty.TRANSCRIPTS_CHANNEL.getValue(server.getId()).map(s -> messageEvent.getChannel().getIdAsString().equals(s)).orElse(false)) {
            List<MessageAttachment> attachments = messageEvent.getMessageAttachments();

            if (attachments.size() != 1) {
                return;
            }

            MessageAttachment attachment = attachments.get(0);

            byte[] attachmentContent = attachment.asByteArray().join();

            String content = new String(attachmentContent, StandardCharsets.UTF_8);

            Optional<Long> channelId = content.lines()
                    .limit(7)
                    .map(String::strip)
                    .map(CHANNEL_FROM_TRANSCRIPT::matcher)
                    .filter(Matcher::find)
                    .map(matcher -> matcher.group("channel"))
                    .map(Long::parseLong)
                    .findFirst();

            String attachmentLink = null;

            Optional<TextChannel> approvingChannel = ServerProperty.LOG_APPROVING_CHANNEL.getValue(server.getId())
                    .flatMap(s -> messageEvent.getApi().getTextChannelById(s));

            for (CarryQueueModel queueModel : QueueConnection.getInstance()
                    .getCarryQueuesByQueueStep(QueueStep.TRANSCRIPT)
                    .orElse(new HashSet<>())
                    .stream().filter(carryQueueModel -> channelId.map(aLong -> aLong.equals(carryQueueModel.getRelationId())).orElse(false))
                    .toList()) {
                if (attachmentLink == null) {
                    Optional<String> attachmentUrl = ContentConnection.getInstance().uploadFile(attachmentContent, "{uuid}.html");

                    if (attachmentUrl.isEmpty()) {
                        logger.error("Couldn't upload content of attachment on message {}.", messageEvent.getMessage().getId());
                        return;
                    }

                    attachmentLink = ContentConnection.getInstance().getCdnUrl(attachmentUrl.get()).toString();
                }

                queueModel.setAttachmentLink(attachmentLink);

                CarryQueueUpdateModel updateModel = new CarryQueueUpdateModel()
                        .setAttachmentLink(attachmentLink);

                if ((queueModel.getAmount() >= APPROVE_AMOUNT_THRESHOLD
                        || queueModel.calculateScore() >= APPROVE_SCORE_THRESHOLD)
                        && approvingChannel.isPresent()) {
                    Message message = approvingChannel.get()
                            .sendMessage(
                                    ApplicationService.getInstance()
                                            .loadEmbedFromCarryQueue(queueModel)
                                            .setTitle("Accept carry-log?")
                                            .setColor(EmbedColor.DEFAULT.getColor()),
                                    ActionRow.of(org.javacord.api.entity.message.component.Button.success(
                                            "accept_log",
                                            "Accept"), Button.danger("deny", "Deny"))
                            ).join();

                    updateModel.setQueueStep(QueueStep.APPROVING)
                            .setRelationId(message.getId());

                    QueueConnection.getInstance().updateQueue(queueModel.getId(), updateModel);
                } else {
                    Long updatedScore = QueueConnection.getInstance()
                            .logQueue(queueModel.getId(), updateModel)
                            .stream()
                            .map(LoggedCarryModel::scoreModels)
                            .flatMap(Collection::stream)
                            .filter(scoreModel -> scoreModel.getScoreType().equals(ScoreType.DEFAULT))
                            .findFirst()
                            .map(ScoreModel::getScoreAmount)
                            .orElseGet(() -> ScoreConnection.getInstance(queueModel.getCarryType())
                                    .getScore(queueModel.getCarrier().getId())
                                    .map(ScoreModel::getScoreAmount)
                                    .orElse(0L));

                    User carrier = messageEvent.getApi().getUserById(queueModel.getCarrier().getId()).join();

                    if (carrier != null && carrier.openPrivateChannel().join() != null) {
                        carrier.openPrivateChannel().join();
                        Optional<PrivateChannel> privateChannelOptional = carrier.getPrivateChannel();
                        long gainedScore = queueModel.calculateScore();

                        try {
                            privateChannelOptional.ifPresent(privateChannel -> privateChannel
                                    .sendMessage("Your carry was logged!\n\n" +
                                                    "**Score gained:** " + gainedScore +
                                                    "\n**Your Updated Score:** " + updatedScore,
                                            ApplicationService.getInstance()
                                                    .loadEmbedFromCarryQueue(queueModel)
                                                    .setTitle("Information")
                                                    .setColor(EmbedColor.DEFAULT.getColor())).join());
                        }
                        catch (CompletionException completionException) {
                            logger.error("Error when sending log message.", completionException);
                        }
                    }

                    Optional<ServerTextChannel> logChannel = queueModel.getCarryTier()
                            .getCarryType()
                            .getLogChannel()
                            .flatMap(server::getTextChannelById);

                    if (logChannel.isPresent()) {
                        logger.debug("Carry logged: {}",
                                DungeonHubService.getInstance().getGson().toJson(queueModel));

                        logChannel.get().sendMessage(
                                ApplicationService.getInstance()
                                        .getEmbed(queueModel.getTime())
                                        .setTitle("Carry accepted.")
                                        .setColor(EmbedColor.POSITIVE.getColor())
                                        .addInlineField("Number of carries",
                                                String.valueOf(queueModel.getAmount()))
                                        .addInlineField("Type of carry",
                                                queueModel.getCarryTier().getDisplayName() + " - " + queueModel.getCarryDifficulty().getDisplayName())
                                        .addInlineField("Player",
                                                messageEvent.getApi().getUserById(queueModel.getPlayer().getId()).join().getMentionTag())
                                        .addInlineField("Carrier",
                                                messageEvent.getApi().getUserById(queueModel.getCarrier().getId()).join().getMentionTag())
                                        .addInlineField("Transcript-Link", "[Click to open]" +
                                                "(" + queueModel.getAttachmentLink() + ")"));
                    }

                    QueueConnection.getInstance().deleteQueue(queueModel.getId());
                }
            }

            LeaderboardService.getInstance().refreshLeaderboard();
        }
    }
}

File: src/main/kotlin/me/taubsie/dungeonhub/application/listener/MessageListener.kt
package me.taubsie.dungeonhub.application.listener

import dev.kord.common.entity.ButtonStyle
import dev.kord.common.entity.Snowflake
import dev.kord.core.behavior.channel.MessageChannelBehavior
import dev.kord.core.behavior.channel.asChannelOfOrNull
import dev.kord.core.behavior.channel.createEmbed
import dev.kord.core.behavior.channel.createMessage
import dev.kord.core.behavior.getChannelOfOrNull
import dev.kord.core.entity.Guild
import dev.kord.core.entity.Message
import dev.kord.core.entity.channel.CategorizableChannel
import dev.kord.core.entity.channel.DmChannel
import dev.kord.core.entity.channel.GuildMessageChannel
import dev.kord.core.entity.channel.TextChannel
import dev.kord.core.event.message.MessageCreateEvent
import dev.kord.core.event.message.MessageUpdateEvent
import dev.kord.rest.builder.message.actionRow
import dev.kordex.core.components.components
import dev.kordex.core.components.disabledButton
import dev.kordex.core.components.ephemeralButton
import dev.kordex.core.components.linkButton
import dev.kordex.core.extensions.Extension
import dev.kordex.core.extensions.event
import dev.kordex.core.i18n.toKey
import dev.kordex.core.utils.addReaction
import dev.kordex.core.utils.dm
import dev.kordex.core.utils.respond
import kotlinx.coroutines.flow.count
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.reduce
import kotlinx.coroutines.runBlocking
import kotlinx.datetime.Clock
import kotlinx.datetime.toKotlinInstant
import me.taubsie.dungeonhub.application.config.ConfigProperty
import me.taubsie.dungeonhub.application.connection.DiscordConnection
import me.taubsie.dungeonhub.application.enums.EmbedColor
import me.taubsie.dungeonhub.application.enums.ServerProperty
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionException
import me.taubsie.dungeonhub.application.exceptions.FailedToLoadEmbedException
import me.taubsie.dungeonhub.application.exceptions.PlayerNotFoundWarning
import me.taubsie.dungeonhub.application.loader.LoadExtension
import me.taubsie.dungeonhub.application.service.ApplicationService
import me.taubsie.dungeonhub.application.service.LeaderboardService
import me.taubsie.dungeonhub.application.service.color
import net.dungeonhub.connection.*
import net.dungeonhub.enums.QueueStep
import net.dungeonhub.enums.ScoreType
import net.dungeonhub.model.carry_queue.CarryQueueModel
import net.dungeonhub.model.score.ScoreModel
import net.dungeonhub.mojang.connection.MojangConnection
import net.dungeonhub.service.MoshiService
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.Request
import org.slf4j.LoggerFactory
import java.nio.charset.StandardCharsets
import java.time.Duration
import java.util.*
import java.util.regex.Matcher
import java.util.regex.Pattern
import kotlin.concurrent.thread
import kotlin.time.toKotlinDuration

@LoadExtension
class MessageListener : Extension() {
    companion object {
        private const val APPROVE_AMOUNT_THRESHOLD: Long = 11
        private const val APPROVE_SCORE_THRESHOLD: Long = 34

        private val CHANNEL_FROM_TRANSCRIPT: Pattern = Pattern.compile("^\\s*Channel: [^(]*\\((?<channel>\\d*)\\)")

        private val logger = LoggerFactory.getLogger(MessageListener::class.java)
    }

    override val name = "message-listener"

    override suspend fun setup() {
        event<MessageCreateEvent> {
            action {
                thread {
                    runBlocking {
                        addReactionToPets(event)

                        logTicket(event)

                        loadSkycryptFromTicket(event)
                    }
                }
            }
        }

        event<MessageUpdateEvent> {
            action {
                thread {
                    runBlocking {
                        logTicket(event)
                    }
                }
            }
        }
    }

    private suspend fun logTicket(event: MessageCreateEvent) {
        logTicket(event.message, event.getGuildOrNull())
    }

    private suspend fun logTicket(event: MessageUpdateEvent) {
        logTicket(
            event.message.asMessage(),
            event.message.getChannelOrNull()?.asChannelOfOrNull<GuildMessageChannel>()?.getGuildOrNull()
        )
    }

    @Synchronized
    private fun logTicket(message: Message, server: Guild?) {
        if (server == null) {
            return
        }

        if (ServerProperty.TRANSCRIPTS_CHANNEL.getValue(server.id.value.toLong()).map { s: String ->
                message.channelId.toString() == s
            }.orElse(false)) {
            val attachments = message.attachments

            if (attachments.size != 1) {
                return
            }

            val attachment = attachments.first()

            val attachmentRequest = Request.Builder().url(attachment.url.toHttpUrl()).build()

            val attachmentData: ByteArray = DungeonHubConnection.executeRawRequest(attachmentRequest)?.result
                ?: throw CommandExecutionException("Couldn't read file data.")

            val content = String(attachmentData, StandardCharsets.UTF_8)

            val channelId: Optional<Long> = content.lines().stream()
                .limit(7)
                .map { obj: String -> obj.trim() }
                .map { input: String ->
                    CHANNEL_FROM_TRANSCRIPT.matcher(
                        input
                    )
                }
                .filter { obj: Matcher -> obj.find() }
                .map { matcher: Matcher ->
                    matcher.group(
                        "channel"
                    )
                }
                .map { s: String -> s.toLong() }
                .findFirst()

            var attachmentLink: String? = null

            val approvingChannel = ServerProperty.LOG_APPROVING_CHANNEL.getValue(server.id.value.toLong())
                .map { s: String ->
                    runBlocking { DiscordConnection.bot?.kordRef?.getChannelOf<TextChannel>(Snowflake(s)) }
                }

            for (queueModel in (QueueConnection.getCarryQueuesByQueueStep(QueueStep.Transcript) ?: HashSet())
                .filter { carryQueueModel: CarryQueueModel ->
                    channelId.map { aLong: Long -> aLong == carryQueueModel.relationId }
                        .orElse(false)
                }) {
                if (attachmentLink == null) {
                    val attachmentUrl = ContentConnection.uploadFile(attachmentData, "{uuid}.html")

                    if (attachmentUrl == null) {
                        logger.error(
                            "Couldn't upload content of attachment on message {}.",
                            message.id
                        )
                        return
                    }

                    attachmentLink = ContentConnection.getCdnUrl(attachmentUrl).toString()

                    thread(start = true) {
                        runBlocking {
                            message.respond {
                                this.content = attachmentLink
                            }
                        }
                    }
                }

                val firstUpdateModel = queueModel.getUpdateModel()
                firstUpdateModel.attachmentLink = attachmentLink

                val updatedModel = QueueConnection.updateQueue(queueModel.id, firstUpdateModel) ?: queueModel

                if ((updatedModel.amount >= APPROVE_AMOUNT_THRESHOLD
                            || updatedModel.calculateScore() >= APPROVE_SCORE_THRESHOLD)
                    && approvingChannel.isPresent
                ) {
                    runBlocking {
                        val createdMessage = approvingChannel.get()
                            .createMessage {
                                val embed = ApplicationService.loadEmbedFromCarryQueue(updatedModel)
                                embed.title = "Accept carry-log?"
                                embed.color = EmbedColor.Default.color

                                embeds = mutableListOf(embed)

                                actionRow {
                                    interactionButton(ButtonStyle.Success, "accept_log") {
                                        label = "Accept"
                                    }

                                    interactionButton(ButtonStyle.Danger, "deny") {
                                        label = "Deny"
                                    }
                                }
                            }

                        thread(start = true) {
                            runBlocking {
                                DiscordConnection.bot?.kordRef
                                    ?.getUser(Snowflake(updatedModel.carrier.id))
                                    ?.dm {
                                        val embed = ApplicationService.embed
                                        embed.color(EmbedColor.Information)
                                        embed.title = "Approval needed"
                                        embed.description =
                                            "Due to the high number of score (${updatedModel.calculateScore()}) or carries (${updatedModel.amount}), your ${updatedModel.carryTier.displayName} - ${updatedModel.carryDifficulty.displayName} log request has to be manually approved by our server's staff team\n" +
                                                    "You will be notified here once it was approved or denied."

                                        embeds = mutableListOf(embed)
                                    }
                            }
                        }

                        val secondUpdateModel = updatedModel.getUpdateModel()

                        secondUpdateModel.queueStep = QueueStep.Approving
                        secondUpdateModel.relationId = createdMessage.id.value.toLong()

                        QueueConnection.updateQueue(updatedModel.id, secondUpdateModel)
                    }
                } else {
                    val updatedScore = QueueConnection.logQueue(updatedModel.id, firstUpdateModel)
                        ?.scoreModels
                        ?.firstOrNull { scoreModel: ScoreModel -> scoreModel.scoreType == ScoreType.Default }
                        ?.scoreAmount
                        ?: (ScoreConnection[updatedModel.carryType].getScore(updatedModel.carrier.id)?.scoreAmount ?: 0)

                    runBlocking {
                        val carrier = DiscordConnection.bot?.kordRef?.getUser(Snowflake(updatedModel.carrier.id))

                        if (carrier != null) {
                            carrier.dm {
                                this.content = "Your carry was logged!\n\n" +
                                        "**Your Updated Score:** $updatedScore"

                                val embed = ApplicationService.loadEmbedFromCarryQueue(updatedModel)
                                embed.title = "Information"
                                embed.color = EmbedColor.Default.color

                                embeds = mutableListOf(embed)
                            }

                            val logChannel = updatedModel.carryTier
                                .carryType
                                .logChannel
                                ?.let { id: Long ->
                                    runBlocking { server.getChannelOfOrNull<GuildMessageChannel>(Snowflake(id)) }
                                }

                            if (logChannel != null) {
                                logger.debug(
                                    "Carry logged: {}",
                                    MoshiService.moshi.adapter(CarryQueueModel::class.java).toJson(updatedModel)
                                )

                                logChannel.createMessage {
                                    val embed = ApplicationService.getEmbed(
                                        updatedModel.time?.toKotlinInstant() ?: Clock.System.now()
                                    )
                                    embed.title = "Carry accepted."
                                    embed.color = EmbedColor.Positive.color
                                    embed.field("Number of carries", true) { updatedModel.amount.toString() }
                                    embed.field("Type of carry", true) {
                                        "${updatedModel.carryTier.displayName} - ${updatedModel.carryDifficulty.displayName}"
                                    }
                                    embed.field("Player", true) {
                                        "<@${updatedModel.player.id}>"
                                    }
                                    embed.field("Carrier", true) {
                                        "<@${updatedModel.carrier.id}>"
                                    }
                                    embed.field("Transcript-Link", true) {
                                        "[Click to open](${updatedModel.attachmentLink})"
                                    }

                                    embeds = mutableListOf(embed)
                                }
                            }

                            QueueConnection.deleteQueue(updatedModel.id)
                        }
                    }
                }

                LeaderboardService.refreshLeaderboard()
            }
        }
    }

    private suspend fun addReactionToPets(event: MessageCreateEvent) {
        if (event.guildId != null) {
            val serverId = event.guildId!!
            val channelId = event.message.channel.id
            if ((serverId.value.toLong() == 1023684107877761196L && channelId.value.toLong() == 1220895875102937098L)
                || (serverId.value.toLong() == 693263712626278553L && channelId.value.toLong() == 1219427157655289908L)
            ) {
                if (event.message.attachments.isEmpty() && event.message.author?.isBot == false &&
                    event.message.embeds.stream()
                        .map { embed -> embed.thumbnail }
                        .filter { it != null }
                        .findFirst().isEmpty
                ) {
                    return
                }

                val emoji: String = getRandomEmoji()

                event.message.addReaction(emoji)
            }
        }
    }

    private fun getEmojiPool(): List<String> {
        return listOf(
            "Woah:1220111116651204608",
            "woah:1220111081150615572",
            "girlwow:1220111157742800956",
            "catelove:1204407157848678430",
            "ZTcool:1204406493353353256",
            "pepega:697756021048868894",
            "smikecate:1204406375791333426",
            "poggorfish:694270485613117480"
        )
    }

    private fun getRandomEmoji(): String {
        val bound = 101.0

        val emojiPool = getEmojiPool()

        val random = Random().nextInt(bound.toInt())

        var emoji = emojiPool[(random * (emojiPool.size / bound)).toInt()]

        if (random == 100) {
            emoji = "swag:708383726370947132"
        }

        return emoji
    }

    private suspend fun loadSkycryptFromTicket(event: MessageCreateEvent) {
        if (event.guildId == null || event.message.channel is DmChannel) {
            return
        }

        val categoryId = event.message.channel.asChannelOfOrNull<CategorizableChannel>()?.categoryId?.value?.toLong()

        if (categoryId != null && !listOf(
                796769131880906782,
                1172688079262330900,
                1168970186666283148,
                992922801075912764,
                992922867857641594,
                1112458713958195250,
                842840550704939053,
                805833601748828200,
                805833672678309908,
                805833843633815664,
                805834037108670464,
                1360864176238493767,
                1360489146321338461,
                1360579229221130421,
                1360575091640897546,
                1364432423345061928,
                1360860377315020941
            ).contains(categoryId)
        ) {
            return
        }

        if (event.message.channel.getMessagesAfter(Snowflake.min, 5).count() != 1) {
            return
        }

        val firstMessage = event.message.channel.messages.reduce { _, message2 -> message2 }

        if (firstMessage.mentionedUsers.count() != 1 || firstMessage.author?.isBot == false) {
            return
        }

        val user = firstMessage.mentionedUsers.first()

        val ign = DiscordUserConnection.getLinkedById(user.id.value.toLong())
            ?.minecraftId
            ?.let { MojangConnection.getNameByUUID(it) }
            ?: return

        sendPlayerDataEmbed(ign, event.message.channel, user.id.value.toLong())
    }

    //TODO threads threads threads
    private suspend fun sendPlayerDataEmbed(ign: String, channel: MessageChannelBehavior, discordId: Long? = null) {
        try {
            channel.createMessage {
                embeds = mutableListOf(ApplicationService.getPlayerDataEmbed(ign, discordId))

                components {
                    linkButton {
                        label = "SkyCrypt".toKey()
                        url = ConfigProperty.SKYCRYPT_API_URL.value + "stats/" + ign
                    }
                }
            }
        } catch (playerNotFoundWarning: PlayerNotFoundWarning) {
            //TODO load scammer data from discord?

            channel.createEmbed { ApplicationService.getErrorEmbed(playerNotFoundWarning) }
        } catch (failedToLoadEmbedException: FailedToLoadEmbedException) {
            channel.createMessage {
                embeds = mutableListOf(failedToLoadEmbedException.embed)

                components(Duration.ofMinutes(5).toKotlinDuration()) {
                    linkButton {
                        label = "SkyCrypt".toKey()
                        url = ConfigProperty.SKYCRYPT_API_URL.value + "stats/" + ign
                    }

                    ephemeralButton {
                        style = ButtonStyle.Secondary
                        label = "Reload".toKey()
                        id = "reload_playerdata"

                        action {
                            edit {
                                components {
                                    linkButton {
                                        label = "SkyCrypt".toKey()
                                        url = ConfigProperty.SKYCRYPT_API_URL.value + "stats/" + ign
                                    }

                                    disabledButton {
                                        style = ButtonStyle.Secondary
                                        label = "Reload".toKey()
                                        id = "reload_playerdata"
                                    }
                                }

                                val embed = ApplicationService.embed
                                embed.description = "Loading..."

                                embeds = mutableListOf(embed)

                                thread {
                                    runBlocking {
                                        components {
                                            embeds = mutableListOf(
                                                try {
                                                    ApplicationService.getPlayerDataEmbed(ign, null)
                                                } catch (failedToLoadAgain: FailedToLoadEmbedException) {
                                                    failedToLoadAgain.embed
                                                }
                                            )
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
